File: libhwio/src/LED.cpp
#include <stdexcept>
#include <ece643/libhwio/LED.hpp>
#include <ece643/libhwio/MMap.hpp>

using namespace std;
using namespace ece643::libhwio;

LED::LED(MMap &mmap) noexcept : mmap(&mmap) {
}

void LED::set(int leds, int hours, int minutes, bool pm) {
    if ((leds & ~0x3FF) || hours < 0 || hours >= 12 || minutes < 0 || minutes >= 60) {
        throw invalid_argument("LED::set");
    }
    (*mmap)[0x200020] = leds | (minutes << 16) | (hours << 24) | (pm ? (1 << 28) : 0);
}

File: sim/src/main/kotlin/ece643/libhwio/LED.kt
package ece643.libhwio

import ece643.sim.LEDView
import ece643.sim.SevenSegView

class LED(private val view: Array<LEDView>, private val sevenseg: SevenSegView) {
    fun set(leds: Int, hours: Int, minutes: Int, pm: Boolean) {
        try {
            for (i in 0 until 10) {
                view[i].active = leds.and(0x200.shr(i)) != 0
            }
            sevenseg.update(hours, minutes, pm)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
